src/components/calendar.rs

```diff
@@ -1,28 +1,263 @@
+use std::fmt::Display;
+
 use maud::{html, Markup};
+use tracing::debug;
+
+use crate::scraper::{Day, Days, MeetingTime, Section};
+
+pub struct HSL {
+    h: u16,
+    s: u8,
+    l: u8,
+}
+
+impl From<u64> for HSL {
+    fn from(value: u64) -> Self {
+        HSL {
+            h: u16::try_from(value % 360).unwrap(),
+            s: 50,
+            l: 25,
+        }
+    }
+}
+
+impl Display for HSL {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "hsl({},{},{})", self.h, self.s, self.l)
+    }
+}
+
+struct RenderableMeeting {
+    top_percent: f32,
+    bottom_percent: f32,
+    title: String,
+    section: String,
+    color: HSL,
+    enr: u32,
+    enr_cap: u32,
+    wl: u32,
+    wl_cap: u32,
+    border: String,
+}
 
-pub fn render() -> Markup {
-    // generates a vec of 8.5->23 in 0.5 increments. this will later be generated
-    // dynamically based on the earliest and latest start time in a person's schedule
-    let timeslots: Vec<f64> = Vec::from_iter((17..46).map(|n| n as f64 * 0.5));
+fn render_section_cards(earliest: i8, latest: i8, sec: &Section) -> Markup {
+    let renderable_meetings: Vec<RenderableMeeting> = sec
+        .meeting_times
+        .iter()
+        .map(|mt| {
+            let title: String = format!("{}{}", sec.subject_code, sec.course_code);
+            let st = match mt.start_time {
+                None => 8.5,
+                Some(t) => t.hour() as f32 + t.minute() as f32 / 60.0,
+            };
+            let et = match mt.end_time {
+                None => 23.5,
+                Some(t) => t.hour() as f32 + t.minute() as f32 / 60.0,
+            };
 
-    let days: Vec<&str> = vec!["monday", "tuesday", "wednesday", "thursday", "friday"];
+            let full = sec.enrollment >= sec.enrollment_capacity;
+            let border = if full {
+                "border-2 border-red-800".to_string()
+            } else {
+                "p-2".to_string()
+            };
+            RenderableMeeting {
+                top_percent: (st - f32::from(earliest)) / f32::from(latest - earliest) * 100.0,
+                bottom_percent: (f32::from(latest) - et) / f32::from(latest - earliest) * 100.0,
+                title,
+                section: sec.sequence_code.to_string(),
+                color: HSL::from(sec.crn),
+                enr: sec.enrollment,
+                enr_cap: sec.enrollment_capacity,
+                wl: sec.waitlist,
+                wl_cap: sec.waitlist_capacity,
+                border,
+            }
```

Comment:
Why not just move the `html!` in here and return `Markup` instead of a custom type? Might help to move the view closer to the variables, that way we can delete the struct type.